// @apple/app-store-server-library/dist/tests/util.js
"use strict";
// Copyright (c) 2023 Apple Inc. Licensed under MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.readFile = readFile;
exports.readBytes = readBytes;
exports.getSignedPayloadVerifier = getSignedPayloadVerifier;
exports.getSignedPayloadVerifierWithDefaultAppAppleId = getSignedPayloadVerifierWithDefaultAppAppleId;
exports.getDefaultSignedPayloadVerifier = getDefaultSignedPayloadVerifier;
exports.createSignedDataFromJson = createSignedDataFromJson;
const fs = require("fs");
const Environment_1 = require("../models/Environment");
const jws_verification_1 = require("../jws_verification");
const crypto_1 = require("crypto");
const jsonwebtoken = require("jsonwebtoken");
function readFile(path) {
    return fs.readFileSync(path, {
        encoding: 'utf8'
    });
}
function readBytes(path) {
    return fs.readFileSync(path);
}
function getSignedPayloadVerifier(environment, bundleId, appAppleId) {
    return new jws_verification_1.SignedDataVerifier([readBytes('tests/resources/certs/testCA.der')], false, environment, bundleId, appAppleId);
}
function getSignedPayloadVerifierWithDefaultAppAppleId(environment, bundleId) {
    return getSignedPayloadVerifier(environment, bundleId, 1234);
}
function getDefaultSignedPayloadVerifier() {
    return getSignedPayloadVerifierWithDefaultAppAppleId(Environment_1.Environment.LOCAL_TESTING, "com.example");
}
function createSignedDataFromJson(path) {
    const fileContents = readFile(path);
    const keyPairOptions = {
        namedCurve: 'prime256v1',
        publicKeyEncoding: {
            type: 'spki',
            format: 'pem'
        },
        privateKeyEncoding: {
            type: 'pkcs8',
            format: 'pem'
        }
    };
    const keypair = (0, crypto_1.generateKeyPairSync)("ec", keyPairOptions);
    const privateKey = keypair.privateKey;
    return jsonwebtoken.sign(fileContents, privateKey, { algorithm: 'ES256' });
}
